parity-crypto/src/secp256k1_alt.rs

```diff
@@ -0,0 +1,411 @@
+	}
+
+
+	/// deprecated, we rather not expose Rng trait, use `keypair_from_slice` instead.
+	/// The intent is to avoid depending on `Rng` trait.
+	fn generate_keypair(r: &mut impl Rng) -> (Self::SecretKey, Self::PublicKey) {
+		let secret_key = random_sec(r);
+		let public_key = PublicKeyInner::from_secret_key(&secret_key.0);
+		(secret_key, PublicKey::new(public_key))
+	}
+
+	/// create a key pair from byte value of the secret key, the calling function is responsible for
+	/// erasing the input of memory.
+	fn keypair_from_slice(sk_bytes: &[u8]) -> Result<(Self::SecretKey, Self::PublicKey), Error> {
+		assert!(sk_bytes.len() == SECRET_SIZE);
+		let secret_key = Self::secret_from_slice(sk_bytes)?;
+		let public_key = Self::public_from_secret(&secret_key)?;
+		Ok((secret_key, public_key))
+	}
+
+	fn public_from_secret(s: &Self::SecretKey) -> Result<Self::PublicKey, Error> {
+		Ok(PublicKey::new(PublicKeyInner::from_secret_key(&s.0)))
+	}
+
+	/// using a shortened 64bit public key as input
+	fn public_from_slice(public_sec_raw: &[u8]) -> Result<Self::PublicKey, Error> {
+
+		let pdata = {
+			let mut temp = [4u8; PUB_SIZE + 1];
+			(&mut temp[1..PUB_SIZE + 1]).copy_from_slice(&public_sec_raw[0..PUB_SIZE]);
+			temp
+		};
+		Ok(PublicKey::new(PublicKeyInner::parse(&pdata)?))
+	}
+
+	fn secret_from_slice(secret: &[u8]) -> Result<Self::SecretKey, Error> {
+		let mut buf = [0;32];
+		buf[..].copy_from_slice(&secret[..SECRET_SIZE]);
+		let res = SecretKey::new(SecretKeyInner::parse(&buf)?);
+		Clear::clear(&mut buf);
+		Ok(res)
+	}
+
+
+}
+
+impl PublicKeyTrait for PublicKey {
+	type VecRepr = Vec<u8>;
+	type CompVecRepr = Vec<u8>;
+
+	fn to_vec(&self) -> Self::VecRepr {
+		self.0.serialize()[1..PUB_SIZE + 1].to_vec()
+	}
+
+	/// Should move to another trait.
+	fn to_compressed_vec(&self) -> Self::CompVecRepr {
+		self.0.serialize_compressed().to_vec()
+	}
+
+	fn verify(&self, signature: &[u8], message: &[u8]) -> Result<bool, Error> {
+		let mut buf = [0;32];
+		buf.copy_from_slice(&message[..]); // panic on incorrect message size
+		let message = Message::parse(&buf);
+		let mut buf = [0;64];
+		buf.copy_from_slice(&signature[..64]); // panic on incorrect signature size
+		let signature = Signature::parse(&buf);
+
+		Ok(secp256k1::verify(&message, &signature, &self.0))
+	}
+
+}
+
+impl SecretKeyTrait for SecretKey {
+	type VecRepr = ClearOnDrop<Vec<u8>>;
+
+	fn to_vec(&self) -> Self::VecRepr {
+		ClearOnDrop::new(self.0.serialize().to_vec())
+	}
+
+	fn sign(&self, message: &[u8]) -> Result<Vec<u8>, Error> {
+
+	 	let mut buf = [0;32];
+		buf.copy_from_slice(&message[..]); // panic on incorrect message size
+		let message = Message::parse(&buf);
+		let (sig, rec_id) = secp256k1::sign(&message, &self.0)?;
+		let mut data_arr = vec![0; 65];
+		data_arr[0..64].copy_from_slice(&sig.serialize());
+		data_arr[64] = rec_id.serialize();
+		Ok(data_arr)
+	}
+
+}
+
+/// random secret key for rand 0.5
+fn random_sec<R: Rng>(rng: &mut R) -> SecretKey {
```

Comment:
Does `SecretKey::random` work for this purpose?

Reply:
I needed external rng source for wasm32 (not using emscripten), but this is not good either the dependency on Rng trait can be avoid through only exposing `secret_from_slice`. I think I should totally remove `generate_keypair` and this function.